homeworks/03-data-structures/src/binary-tree.ts

```diff
@@ -1 +1,71 @@
-export const example = "test"
+import {TraverseType, TreeNode} from "./tree-common"
+
+export interface BinaryTree<T> {
+    root(): TreeNode<T>
+    setTree(value: TreeNode<T>): this
+    traverse(traverseType: TraverseType): readonly T[]
+    getColumn(columnOrder: number): readonly T[]
+}
+
+export class BinaryTreeImpl<T> implements BinaryTree<T> {
+    private _root: TreeNode<T>
+
+    constructor(tree: TreeNode<T>) {
+        this._root = tree
+    }
+
+    getColumn(columnOrder: number): readonly T[] {
+        function collect(current: number, node: TreeNode<T> | null): readonly T[] {
+            if (node === null) {
+                return []
+            } else {
+                const self = current === columnOrder ? [node.value] : []
+                const left = collect(current - 1, node.left)
+                const right = collect(current + 1, node.right)
+                return [...self, ...left, ...right]
+            }
+        }
+
+        return collect(0, this.root())
+    }
+
+    /** @deprecated Mutates the internal state, avoid. */
```

Comment:
Not sure why this should be deprecated and avoided?
The idea of `setTree` function is to replace the whole tree with new one.

Reply:
Yes, that's the idea. I disagree with this idea though, I think it's bad design.